File: src/nodes/_typeof.rs
use apica_common::bytecodes::ApicaTypeBytecode;
use crate::utils::position::Position;

#[derive(Clone)]
pub struct NodeTypeof {
    position: Position,
    value_kind: ApicaTypeBytecode,
}

impl NodeTypeof {
    pub fn init(position: Position, value_kind: ApicaTypeBytecode) -> NodeTypeof {
        NodeTypeof { position, value_kind }
    }

    pub fn get_position(&self) -> &Position {
        &self.position
    }

    pub fn show(&self, indent: &mut String, end: char) {
        print!("{indent}NodeTypeof(type: {:?}){end}", self.value_kind);
    }
    
    pub fn get_value_kind(&self) -> ApicaTypeBytecode {
        self.value_kind
    }
}

File: src/nodes/typeof.cpp
#include "nodes/typeof.hpp"
#include "core/emitter.hpp"
#include "utils/builtins.hpp"
#include "utils/diagnostic_bag.hpp"
#include "utils/errors.hpp"
#include <iostream>

using namespace nodes;

NodeTypeof::NodeTypeof(const utils::Position &position, const std::string &vtype)
    : Node(position), value_type(vtype) {

}

void NodeTypeof::show(std::string &indent, char end) const {
    std::cout << indent << "NodeTypeof(type: " << this->value_type << ')' << end;
}

NodeKind NodeTypeof::getKind() const {
    return NodeKind::Typeof;
}

void NodeTypeof::emit(core::Emitter &emitter) const {
    auto declared_type = utils::APICA_TYPES.find(this->value_type);
    if (declared_type == utils::APICA_TYPES.end()) {
        utils::DiagnosticBag::getInstance().addDiagnostic(utils::Diagnostic(
            utils::DiagnosticKind::Error,
            std::string(utils::EMT_ERROR_INCORRECT_TYPE),
            this->position
        ));
    } else {
        emitter.writeU64(common::bytecodes::ApicaBytecode::Literal);
        emitter.writeU64(common::bytecodes::ApicaTypeBytecode::Type);
        emitter.writeU64(declared_type->second);
    }
}

void NodeTypeof::setId() {
    
}

utils::OptimizedResult NodeTypeof::optimize(core::Optimizer &) {
    return utils::OptimizedResult(utils::OptimizedFlag::None);
}